### PR TITLE
Use the macro structure spans instead of the invocation

### LINKED ISSUE #42104: Backward spans are displayed strangely
```rust
macro_rules! m {
    ($a:tt $b:tt) => {
        $b $a;
    }
}

fn main() {
    m!(S struct);
}
```

This is a regression in 1.18.0-beta but unclear whether it affects anything in practice. We encountered this in #rust-libs discussing https://github.com/rust-lang/rust/pull/40939.

```
$ rustc +stable --version
rustc 1.17.0 (56124baa9 2017-04-24)
$ cargo +stable build
   Compiling testing v0.1.0
warning: struct is never used: `S`
 --> src/main.rs:8:10
  |
8 |     m!(S struct);
  |     -----^------- in this macro invocation
  |
  = note: #[warn(dead_code)] on by default

    Finished dev [unoptimized + debuginfo] target(s) in 0.34 secs
```

```
$ rustc +beta --version
rustc 1.18.0-beta.2 (f4e8e81c3 2017-05-11)
$ cargo +beta build
   Compiling testing v0.1.0
warning: struct is never used: `S`
 --> src/main.rs:8:10
  |
3 | |         $b $a;
  | |______________^
...
8 |       m!(S struct);
  |  _____-----^-------
  | |     |
  | |     in this macro invocation
  |
  = note: #[warn(dead_code)] on by default

    Finished dev [unoptimized + debuginfo] target(s) in 0.45 secs
```

### PR DESCRIPTION
**Pull request by estebank**
Fix #42104, CC #2887.

### CODE REVIEW

File: src/libsyntax_pos/lib.rs [hunk 1]
```diff
@@ -184,15 +185,32 @@ impl Span {
         result
     }
 
+    pub fn empty_ctxt(&self) -> bool {
```
Comment (petrochenkov) [inline src/libsyntax_pos/lib.rs]: This new function is unused.

File: src/libsyntax_pos/lib.rs [hunk 2]
```diff
@@ -184,15 +185,32 @@ impl Span {
...
+        let lo = if self.lo < end.lo {
+            self.lo
+        } else {
+            end.lo
+        };
+        let hi = if self.hi > end.hi {
+            self.hi
+        } else {
+            end.hi
+        };
```
Comment (petrochenkov) [inline src/libsyntax_pos/lib.rs]: Probably better to use standard `min` and `max` for clarity.

File: src/libsyntax/tokenstream.rs [hunk 3]
```diff
@@ -241,6 +258,21 @@ impl TokenStream {
         }
     }
 
+    pub fn map_pos<F: FnMut(usize, TokenTree) -> TokenTree>(self, mut f: F) -> TokenStream {
```
Comment (petrochenkov) [inline src/libsyntax/tokenstream.rs]: I'd rename this into `enumerate_map` (or `map_enumerated`), to use the same naming conventions as `Iterator`.

File: src/libsyntax/ext/tt/macro_rules.rs [hunk 4]
```diff
@@ -119,8 +119,21 @@ fn generic_extension<'cx>(cx: &'cx mut ExtCtxt,
                     quoted::TokenTree::Delimited(_, ref delimed) => delimed.tts.clone(),
                     _ => cx.span_bug(sp, "malformed macro rhs"),
                 };
+
                 // rhs has holes ( `$id` and `$(...)` that need filled)
-                let tts = transcribe(cx, Some(named_matches), rhs);
+                let mut tts = transcribe(cx, Some(named_matches), rhs.clone());
```
Comment (petrochenkov) [inline src/libsyntax/ext/tt/macro_rules.rs]: Just to be sure, everything is under `Rc` and cloning/mapping is cheap, right?

**Comment by petrochenkov** [comment]
Reviewed.

File: src/libsyntax/ext/tt/macro_rules.rs (hunk 4 above)
Comment (estebank) [inline src/libsyntax/ext/tt/macro_rules.rs]: All I need is the spans after this, so I'm cloning the spans first into their own `Vec`.

File: src/libsyntax/tokenstream.rs (hunk 3 above)
Comment (estebank) [inline src/libsyntax/tokenstream.rs]: Done.

File: src/libsyntax_pos/lib.rs (hunk 1 above)
Comment (estebank) [inline src/libsyntax_pos/lib.rs]: Removed.

File: src/libsyntax_pos/lib.rs (hunk 2 above)
Comment (estebank) [inline src/libsyntax_pos/lib.rs]: Done.

**Pull request merged**